caddyconfig/caddyfile/lexer.go

```diff
@@ -168,3 +169,23 @@ func (l *lexer) next() bool {
 		val = append(val, ch)
 	}
 }
+
+// Tokenize takes bytes as input and lexes it into
+// a list of tokens that can be parsed as a Caddyfile.
+// Also takes a filename to fill the token's File as
+// the source of the tokens, which is important to
+// determine relative paths for `import` directives.
```

Comment:
Add a scary warning that this function is experimental and is not guaranteed to be stable.

Reply:
I don't think it's experimental... It's the exact same code that was in parser.go

Reply:
Whether it's experimental or not, we need to say that it's not stable ("experimental" is a good consistent term) because I wasn't ever expecting to export this API when I designed it.

Reply:
I'm not sure I understand how it wouldn't be stable if it's actually used by the Caddyfile parser itself.

Reply:
I just mean that right now we can change it whenever we want, because we know it's unexported and nobody else is using it. I just try to avoid exporting things.

Reply:
As I mentioned in the other PR, this is not a must-have.
We can give it some time and see how caddy-docker-proxy code will evolve.
And also see if the lexer will change that often in caddy

Reply:
@lucaslorentz Is `Tokenize()` the right function to export? Or do you need env var replacements, I can't remember. If you need env var replacements, maybe allTokens is better: https://github.com/caddyserver/caddy/pull/3549/files#diff-06ecc7b9618d6f991e3d8846d817e0dcR85

Reply:
I'm pretty certain we don't want env var replacement to happen because the step that caddy-docker-proxy takes is just rebuilding a valid Caddyfile from the labels the user inputs. The env vars can be replaced later when Caddy actually runs. If env var replacement was done at that stage, then it would use the env of the controller, which might not be the correct stuff.

Reply:
I don't need env var replacements.

I have my own in memory representation of caddyfile, that doesn't depend plugins structs.
https://github.com/lucaslorentz/caddy-docker-proxy/blob/master/plugin/caddyfile/caddyfile.go

I also have a tiny parser, that parses all tokens into my in memory caddyfile struct:
https://github.com/lucaslorentz/caddy-docker-proxy/blob/master/plugin/caddyfile/marshal.go#L164

Currently, the tokenization is an overlap between caddy implementation and my own implementation. Thus, we're considering exposing that feature.

But, another option, is to expose in caddy a function similar to the existing parseCaddyfile, but that keeps env variables and snippets untouched. Then I could change my parser to read siteBlocks returned by parseCaddyfile and tokens from each siteBlock.

Reply:
I see, okay. Thanks!